.ci/magician/exec/runner.go

```diff
@@ -0,0 +1,98 @@
+package exec
+
+import (
+	"container/list"
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+
+	cp "github.com/otiai10/copy"
+)
+
+type actualRunner struct {
+	curDir   string
+	dirStack *list.List
+}
+
+type Runner interface {
+	GetCurDir() string
+	Copy(src, dest string) error
+	RemoveAll(path string) error
+	PushDir(path string) error
+	PopDir() error
+	WriteFile(name, data string) error
+	Run(name string, args, env []string) (string, error)
+	MustRun(name string, args, env []string) string
+}
+
+func NewRunner() (Runner, error) {
+	wd, err := os.Getwd()
+	if err != nil {
+		return nil, err
+	}
+	return &actualRunner{
+		curDir:   wd,
+		dirStack: list.New(),
+	}, nil
+}
+
+func (ar *actualRunner) GetCurDir() string {
+	return ar.curDir
+}
+
+func (ar *actualRunner) Copy(src, dest string) error {
+	return cp.Copy(src, dest)
+}
+
+func (ar *actualRunner) RemoveAll(path string) error {
+	return os.RemoveAll(path)
+}
+
+// PushDir changes the directory for the runner to the desired path and saves the previous directory in the stack.
+func (ar *actualRunner) PushDir(path string) error {
+	if ar.dirStack == nil {
+		return errors.New("attempted to push dir, but stack was nil")
+	}
+	ar.dirStack.PushFront(ar.curDir)
+	ar.curDir = path
+	return os.Chdir(path)
```

Comment:
I worry about changing the os's cwd then doing operations from there.

It means we cannot use Runners in parallel if they have different working directories. Can we change the os's directory briefly to the new cwd before preforming any operations?

This way we aren't over writing a global state in os.

Reply:
I think that would still not allow us to use runners in parallel safely, since the magician's working directory could still be in an unexpected place if two commands happen to run at the same time.

However, this would only be true if the parallelization is happening via go-routines. If we're running the magician in separate processes their working directories should be specific to the process. At least, that's how it worked on my machine when I tested it.

Reply:
Are we using parallelization today? Is there another workaround so we don't mutate the global os object?

Reply:
I don't think we are using parallelization in a way that would be problematic here, as long as there is only one Runner per process.

If we wanted to have a way of running things in parallel within one process, I think the best way would be to make the runner keep a set of running processes and a working directory for each one. It could then switch to the correct directory for the correct process before running commands.

Reply:
I would say lets not assume a parallelization, but we should support multiple instances or runner. These instances should not share a global state (which they currently do with os). If we can switch to the directory before pushing performing os operations then switch back lets do that. I know this won't cover parallelization but thats fine. We are supporting multiple instances of runners.

Reply:
CMD can use the DIR field you mentioned. IMO we need it to be the case. Object creation should not share state with other instances of the same type.

Reply:
Using the `Dir` field now.